### DISCUSSION TITLE
Is it completely impossible to run SSR queries from within components?

### DISCUSSION THREAD

**Discussion by YCMitch** [vercel/next.js#19293] [Help]
I'm aware we can use `getInitialProps`, or `getServerSideProps` from a page to query things on the server.

In Gatsby (I know, static vs SSR), we can use useStaticQuery, in PHP everything is server-side. In NextJS is it completely impossible to query things within components on the server?

I can't find any documentation around it. If so, why?

**Comment by ljosberinn** [comment]
`useStaticQuery` fetches data at build time, so its comparable to `getStaticProps`

**Comment by ashmore11** [comment]
Hi @MitchEff 

If you want to query data from a component after rendering, you can simply use native API's like [fetch](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) or an abstraction like [Apollo's useQuery](https://www.apollographql.com/docs/react/api/react/hooks/#usequery) if using GraphQL.

**Comment by YCMitch** [comment]
Cheers guys - perhaps I should've been a little clearer, I more meant running a server-side query from within a component. My `useStaticQuery` example was more that you can use it in any component (rather than just page components like `getServerSideProps`). 

useQuery and useSWR are great for the client, but don't render on the server.

**Reply by ashmore11** [reply]
Ah I see. Unfortunately there's no way to query data outside of pages currently AFAIK.

**Reply by ljosberinn** [reply]
both react query and swr are SSR compatible

https://react-query.tanstack.com/docs/guides/ssr
https://github.com/vercel/swr#ssr-with-nextjs

**Reply by YCMitch** [reply]
> both react query and swr are SSR compatible
> 
> https://react-query.tanstack.com/docs/guides/ssr
> https://github.com/vercel/swr#ssr-with-nextjs

They sure are, but it doesn't really help us do SSR queries outside of pages in NextJS, unfortunately.

**Reply by ljosberinn** [reply]
I'm still confused what you're looking for. You initially said

> In NextJS is it completely impossible to query things within components on the server?

then redacted to 

>  I more meant running a server-side query from within a component.

Now you're talking about

> do SSR queries outside of pages in NextJS

Could you give a real world example?

**Reply by YCMitch** [reply]
@ljosberinn, they're all the exact same thing. 

I'd like to run queries inside React components **on the server** (i.e, a server-side query within a component, i.e an SSR query _not just in a page_).

A simple example would be querying for menu items in the header. You can use useSWR, Apollo, whatever, but it won't show up in Google or anything, as it's only done client-side. 

@ashmore11 appears to be correct in that you can't do this in NextJS for whatever reason. I'm aware Gatsby is SSG not SSR, but you can write queries _anywhere_ and they'll be SEO-friendly. In NextJS I guess you need to query absolutely everything from your page template.

**Reply by ashmore11** [reply]
@MitchEff What you could do is query your header from within each page and pass the `pageProps` from a custom `_app` to your Header component.

You could also make the query directly from `_app` using `getInitialProps` but then you’d be opting out of automatic static optimisation for any static page that doesn’t explicitly use `getStaticProps`.

Either approach is not ideal but they will accomplish what you need.

**Reply by ashmore11** [reply]
Also just to add. Next.js uses SSG _*and*_ SSR.

Any page that exports `getStaticProps` will compile to html at build time.

Any page that exports `getServerSideProps` will compile at run time on the server.

That’s what makes Next.js so awesome!

**Reply by ljosberinn** [reply]
> In NextJS I guess you need to query absolutely everything from your page template.

Yup, I don't see how this is possibly a bad thing. It's one of the rare cases where I'd clearly say its useful separation of concerns. Your backend logic has no business in UI logic. It's still in the same _file_, but separated.
